### ISSUE TITLE
W4.1: Refactor trainer to HuggingFace accelerate

### ISSUE THREAD

**Issue by freddysongg** [freddysongg/LLM-W#30]
## PRD

**Problem:** Current trainer uses `trl.SFTTrainer` via subprocess (`backend/app/services/trainer.py`). No distributed-training support. Multi-GPU requires HuggingFace `accelerate` integration.

**Goal:** Refactor trainer entry point to use `accelerate.Accelerator`. Single-GPU loss curve must match pre-refactor within 1e-3. Stdout-JSON event contract for orchestrator preserved.

### Dependency

Add to `backend/pyproject.toml` optional group `training`:

```toml
training = [
    "accelerate>=0.30,<1.0",
    # ... existing
]
```

### Refactor

File: `backend/app/services/trainer.py`.

Current: directly constructs `trl.SFTTrainer`.
New: wrap in `accelerate.Accelerator`.

```python
from accelerate import Accelerator

def train(config: WorkbenchConfig) -> None:
    accelerator = Accelerator()
    model = ...  # as before
    tokenizer = ...
    train_ds = ...

    trainer = SFTTrainer(
        model=model,
        tokenizer=tokenizer,
        train_dataset=train_ds,
        args=training_args,
    )

    # accelerate.prepare handles device placement
    trainer = accelerator.prepare(trainer)

    # Only rank 0 emits stdout JSON events
    if accelerator.is_main_process:
        emit_event({"type": "stage.start", ...})

    trainer.train()

    if accelerator.is_main_process:
        emit_event({"type": "stage.end", ...})
```

### Stdout gating

Critical: existing orchestrator reads JSON events from stdout. In FSDP, N processes would each emit → orchestrator chokes. All event emission gated on `accelerator.is_main_process`.

### Loss parity test

Add `backend/tests/test_accelerate_parity.py`:

1. Run 50 training steps on a small model (e.g., `sshleifer/tiny-gpt2`) using pre-refactor trainer
2. Run 50 training steps using new accelerate wrapper, same seed, same data
3. Assert `abs(loss_pre - loss_post) < 1e-3` at each step

Save side-by-side loss curves to `docs/assets/fsdp-loss-match.png`.

## Acceptance Criteria

- [ ] `accelerate` added to pyproject.toml `training` group with version cap
- [ ] `backend/app/services/trainer.py` uses `Accelerator`
- [ ] Single-GPU mode: loss curve matches pre-refactor within 1e-3 over 50 steps
- [ ] Loss-parity chart committed at `docs/assets/fsdp-loss-match.png`
- [ ] Stdout JSON events only emitted from `is_main_process == True`
- [ ] Orchestrator ingests events unchanged (no parse errors on single-GPU run)
- [ ] Existing training page still starts runs; Runs page still shows live metrics
- [ ] `mypy backend/app` passes

## Dependencies

- **Blocked by:** none
- **Blocks:** #31 (configs reference accelerate), #32 (rank-aware obs relies on this), #33 (atomic checkpointing), #34 (cloud FSDP run)

## References

- Spec: `docs/plans/2026-04-13-llm-w-improvement-plan-v4.md` §W4.1
- Spec risk: §R-13 (stdout contract preservation)
- Existing: `backend/app/services/trainer.py`, `backend/app/services/training_dispatcher.py`, `backend/app/services/orchestrator.py`

**Comment by freddysongg** [comment]
Landed in merge commit 124caf2 (main). Refactor: trainer wrapped with explicit `accelerate.Accelerator()`, kept `trl.SFTTrainer` as training engine (minimum-drift path for stdout contract). Stdout event contract, metric field names (`train_loss`, `tokens_per_second`, `gpu_memory_used_mb`), and per-run checkpoint layout all preserved. Emission gated on main process for future FSDP.

Also folds in #53 — callback stage renamed to `callback_evaluation`, disjoint from reserved stage 11 `evaluation`. Test coverage landed in bf22286 (follow-up): config sets `eval_split: train` and `eval_steps: 2` so `WorkbenchCallback.on_evaluate` actually fires, and `include_num_input_tokens_seen=True` on `SFTConfig` so `tokens_per_second` emission can be asserted.

Known remaining work filed separately:
- Loss-parity chart (`docs/assets/fsdp-loss-match.png`) — bench-machine deliverable requiring compute; AC not yet met
- Accelerate configs (`configs/accelerate_*.yaml`) — tracked in #31
- FSDP wiring — tracked in #32-#34
- Dead `_atomic_checkpoint_write` helper left in place per strict scope
